### PR TITLE
tf.network: check for data availabibility

### PR DESCRIPTION
**Pull request by JackTemaki**
- Adds a DataNotAvailable Exception
 - Aborts construction stack when the output is flagged
   as available_for_inference=False

This is related to issue #372, and approaches the problem by introducing an Exception that is caught when calling the network construction from "register_as_extern_data". In all other cases the Exception will be raised globally.

### CODE REVIEW

**Comment by albertz** [comment]
But as I explained in #372, this approach has the problem that there is no real way that you can make `register_as_extern_data` work.

I think the approach via `_TemplateLayer` or `UnavailableDataLayer` is better, because then `register_as_extern_data` works naturally.

**Comment by JackTemaki** [comment]
I did not really get your point in the issue discussion, but maybe I do now. I think what you mean to say is that if you use any kind of dummy layer which only returns the templated output, the templated output will be added to the extern_data and thus being able to give other layers that depend on the registered data the information that the data is not available.

Also, I did not see that you can use the "output" parameter of the `LayerBase` directly to create an empty layer with the specified output.

File: returnn/tf/network.py [hunk 1]
```diff
@@ -695,6 +695,9 @@ def _create_layer(self, name, layer_class, **layer_desc):
           layer_class.__name__, name, layer_desc)
         output_template.sanity_check(ignore_placeholder=True)  # placeholder might be overwritten later
         output_template_special_axes = output_template.get_special_axes_dict()
+        if output_template.available_for_inference is False and not self.eval_flag:
```
Comment (albertz) [inline returnn/tf/network.py:698]:
```suggestion
        if not output_template.available_for_inference and not self.eval_flag:
```

File: returnn/tf/network.py [hunk 2]
```diff
@@ -695,6 +695,9 @@ def _create_layer(self, name, layer_class, **layer_desc):
...
+          from returnn.tf.layers.base import DataNotAvailableLayer
+          return DataNotAvailableLayer(layer_desc['name'], layer_desc['network'], output_template)
```
Comment (albertz) [inline returnn/tf/network.py:700]: Always use kwargs.

File: returnn/tf/layers/base.py [hunk 3]
```diff
@@ -1344,6 +1344,14 @@ class InternalLayer(LayerBase):
   """
 
 
+class DataNotAvailableLayer(LayerBase):
```
Comment (albertz) [inline returnn/tf/layers/base.py:1347]: It should derive from `InternalLayer`.

File: returnn/tf/layers/base.py [hunk 4]
```diff
@@ -1344,6 +1344,14 @@ class InternalLayer(LayerBase):
   """
 
 
+class DataNotAvailableLayer(InternalLayer):
+  """
+  This is dummy layer that is created when the output template is "not available for inference"
+  The output
+
+  """
```
Comment (albertz) [inline returnn/tf/layers/base.py:1352]: Incomplete documentation?

**Comment by albertz** [comment]
Is this ready? Does it work?

**Comment by JackTemaki** [comment]
Ok if this is the question, then I better add a specific test case.

File: returnn/tf/layers/base.py (hunk 4 above)
Comment (JackTemaki) [reply returnn/tf/layers/base.py:1352]: Ah yes, it seems I stashed a part of it away.

**Comment by albertz** [comment]
I mean, if you verified / tested it for your specific case, we don't necessarily need a test case. But I would not add this without any testing at all.

**Comment by JackTemaki** [comment]
It seems it was a good decision to build a test case anyway. While it does work for my TTS system ("forward"-task), it does not work yet for the test I wrote that is using search.

**Comment by JackTemaki** [comment]
I was not able to get my test case running so far, and I did not even include the "register_as_extern_data" part so far. It seems there is either a bug or I am misunderstanding something in how search should be performed in RETURNN.

**Comment by albertz** [comment]
So probably `classes` ends up (incorrectly) in `used_data_keys`, as this is checked in `check_matched_dataset`. So the question is, why does it end up there. It should not be used (with search).

**Comment by JackTemaki** [comment]
So the `used_data_keys` were correct, but the problem was in the `extra_fetches`. I just blocked by the relevant code parts by checking for `do_eval` now, but we could also add a different parameter. This will at least keep the default behaviour in all cases that do not crash.

**Review by albertz — APPROVED** [review]
Looks ok.

**Pull request merged**